File: 15.py
from collections import defaultdict
import sys
from copy import deepcopy


class Un:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.h = 200
        self.gone = 0
    
    def p(self):
        return (self.x, self.y)

    def attack(self, foes):
        fs = []
        for c in circle(self.p()):
            f = find(c, foes)
            if f is not None:
                fs.append(f)
        if fs:
            f = min(fs, key=lambda u: u.h)
            if foes == elves or len(sys.argv)<2:
                f.h -= 3
            else:
                f.h -= int(sys.argv[1])
            if f.h <= 0:
                foes.remove(f)
            return 1
        return 0

    def go(self, friends, foes):
        if not self.attack(foes):
            step = route((self.x, self.y), to_attack(foes))
            if step is not None:    
                self.x = step[0]
                self.y = step[1]
                self.attack(foes)
        self.gone = 1

def find(p, l):
    for u in l:
        if u.x == p[0] and u.y == p[1]:
            return u

def clear(l):
    for i in l:
        i.gone = 0
            
def circle(p):
    yield (p[0], p[1]-1)
    yield (p[0]-1, p[1])
    yield (p[0]+1, p[1])
    yield (p[0], p[1]+1)

def to_attack(units):
    fin = set()
    for unit in units:
        fin.update(circle(unit.p()))
    return fin

#taken straight from SO like a full blood pajeet
def route(start, points):
    visited = set()
    distances = defaultdict(lambda: float("inf"))
    distances[start] = 0
    to_visit = [start]
    tentative_parents = {}
    while to_visit:
        nx = min(to_visit, key=lambda x: distances[x])
        if nx in points:
            break
        to_visit.remove(nx)
        visited.add(nx)
        for n in circle(nx):
            if n in visited or terrain[n[1]][n[0]] == '#' or\
                            find(n, gobs) or find(n, elves):
                continue
            nd = distances[nx] + 1
            if nd < distances[n]:
                distances[n] = nd
                tentative_parents[n] = nx
                to_visit.append(n)
    else:
        return None
    path = []
    while nx:
        path.append(nx)
        nx = tentative_parents.get(nx)
    return path[-2]

def disp():
    t = deepcopy(terrain)
    for u in elves:
        t[u.y][u.x] = "E"
    for u in gobs:
        t[u.y][u.x] = "G"
    print("\n".join("".join(l) for l in t))

def healthes():
    for e in elves:
        print("E:", e.h)
    for e in gobs:
        print("G:", e.h)

data = open("15-test.txt").read().split("\n")
elves = [Un(x, y) for y, l in enumerate(data) for x, c in enumerate(l) if c == 'E']
gobs =  [Un(x, y) for y, l in enumerate(data) for x, c in enumerate(l) if c == 'G']
terrain = [['.' if c in ".GE" else '#' for c in l] for l in data]

start_len = len(elves)

rounds = 0
while len(elves) and len(gobs):
    disp()
    healthes()
    while 1:
        try:
            y, x = min(( (u.y, u.x) for u in elves+gobs if u.gone == 0))
        except:
            break
        f = find((x, y), elves)
        if f is not None:
            f.go(elves, gobs)
        f = find((x, y), gobs)
        if f is not None:
            f.go(gobs, elves)
    if elves and gobs:
        rounds += 1
    #print(rounds)
    clear(elves)
    clear(gobs)

H = sum((u.h for u in gobs+elves))

print(rounds * H)

if len(elves) == start_len: 
    print("H:", rounds * H, file=sys.stderr)
    exit(0)
    print("elves")
exit(1)

File: 15.c
#include "common.h"

typedef struct Unit {
    int x;
    int y;
    int h;
    int g;
} Unit;

typedef struct Force {
    Unit units[100];
    int n;
} Force;
#define EACH(f, v) for(int v=0; v<(f)->n; v++)

Force elves;
Force gobs;

#define TERR(x, y) buf[(y)*(w+1)+(x)]
char buf[1057];
int w, h;
Unit circle[] = {{0, -1}, {-1, 0}, {+1, 0}, {0, +1}};
int elf_strength;

Unit * find(int x, int y, Force * f){
    EACH(f, i)
        if(f->units[i].x == x && f->units[i].y == y)
            return &f->units[i];
    return NULL;
}

void attack(Unit * un, Force * f){
    Unit * as = NULL;
    Unit * a;
    for(int c=0; c<LEN(circle); c++)
        if((a = find(un->x+circle[c].x, un->y+circle[c].y, f)))
            as = as && as->h<=a->h ? as : a;
    if(!as) return;
    if(f == &elves)
        as->h -= 3;
    else
        as->h -= elf_strength;
    if(as->h <= 0)
        *as = f->units[--f->n];
}

int unit_cmp(Unit * a, Unit * b, void *k){
    struct {int d[w][h];} * j = k;
    return  j->d[b->x][b->y] - j->d[a->x][a->y] ?: 
            b->y - a->y ?:
            b->x - a->x;
}

void go(Unit * un, Force * friends, Force * foes){
    Unit pars[w][h];
    int distances[w][h]; 
    memset(distances, 0x7f, sizeof(distances));
    distances[un->x][un->y] = 0;
    Unit to_visit[w*h];
    to_visit[0] = *un;
    int nto_visit = 1;
    while(nto_visit){
        Unit cur = to_visit[--nto_visit];
        if(find(cur.x, cur.y, foes)){
            Unit step;
            int away = 0;
            while(cur.x != un->x || cur.y != un->y){
                away ++;
                step = cur;
                cur = pars[cur.x][cur.y];
            }
            if(away>1){
                un->x = step.x;
                un->y = step.y;
            }
            if(away<3)
                attack(un, foes);
            break;
        }
        for(int c=0; c<LEN(circle); c++){
            Unit n = {cur.x+circle[c].x, cur.y+circle[c].y};
            if(TERR(n.x, n.y)=='#' || find(n.x, n.y, friends))
                continue;
            int nd = distances[cur.x][cur.y] + 1;
            if(nd < distances[n.x][n.y]){
                distances[n.x][n.y] = nd;
                pars[n.x][n.y] = cur;
                to_visit[nto_visit++] = n;
            }
        }
        qsort_r(to_visit, nto_visit, sizeof(Unit), (int (*)(const void *, const void*, void*))unit_cmp, distances);
    }
    un->g = 1;
}

void disp(void){
    for(int x=0; x<w; x++) for(int y=0; y<h; y++)
        if(TERR(x, y) == 'E' || TERR(x, y) == 'G')
            TERR(x, y) = '.';
    EACH(&elves, e)
        TERR(elves.units[e].x, elves.units[e].y) = 'E';
    EACH(&gobs, g)
        TERR(gobs.units[g].x, gobs.units[g].y) = 'G';
    puts(buf);
}

void healthes(void){
    EACH(&elves, e)
        printf("E: %d\n", elves.units[e].h);
    EACH(&gobs, g)
        printf("G: %d\n", gobs.units[g].h);
}

Unit * _choose(Unit cur, Force * f){
    Unit * res = NULL;
    EACH(f, i) if(!f->units[i].g && (cur.y - f->units[i].y ?: cur.x - f->units[i].x) >= 0){
        cur.x = f->units[i].x;
        cur.y = f->units[i].y;
        res = &f->units[i];
    }
    return res;
}

int choose(void){
    Unit * elf = _choose((Unit){w, h}, &elves),
         * gob = _choose(elf?*elf:(Unit){w, h}, &gobs);
    if(gob)
        go(gob, &gobs, &elves);
    else if(elf)
        go(elf, &elves, &gobs);
    else
        return 0;
    return 1;
}

int main(int argc, char ** argv){
    if(argc>1) elf_strength = atoi(argv[1]);
    else       elf_strength = 3;
    int fd = open("15-test.txt", O_RDONLY);
    int n = read(fd, buf, LEN(buf));
    w = strchr(buf, '\n') - buf;
    h = n / (w+1);
    for(int x=0; x<w; x++) for(int y=0; y<h; y++){
        if(TERR(x, y) == 'E')
            elves.units[elves.n++] = (Unit){x, y, 200, 0};
        if(TERR(x, y) == 'G')
            gobs.units[gobs.n++]   = (Unit){x, y, 200, 0};
    }
    int start_len = elves.n;
    int rounds = -1;
    while(elves.n && gobs.n){
        rounds++;
        disp();
        healthes();
        while(choose());
        EACH(&elves, i) elves.units[i].g = 0;
        EACH(&gobs, i)  gobs.units[i].g = 0;
    }
    int S = 0;
    EACH(&elves, i) S += elves.units[i].h;
    EACH(&gobs, i)  S += gobs.units[i].h;
    printf("H: %d\n", S * rounds);
    if(start_len == elves.n)
        return 0;
    return 1;
}
